framework/src/base/MeshGeneratorSystem.C

```diff
@@ -0,0 +1,496 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MeshGeneratorSystem.h"
+
+#include "MooseApp.h"
+#include "MeshGenerator.h"
+#include "DependencyResolver.h"
+
+MeshGeneratorSystem::MeshGeneratorSystem(MooseApp & app)
+  : PerfGraphInterface(app.perfGraph(), "MeshGeneratorSystem"),
+    ParallelObject(app),
+    _app(app),
+    _popped_final_mesh_generator(false)
+{
+}
+
+void
+MeshGeneratorSystem::addMeshGenerator(const std::string & type,
+                                      const std::string & name,
+                                      const InputParameters & params)
+{
+  mooseAssert(!_mesh_generator_params.count(name), "Already exists");
+  _mesh_generator_params.emplace(
+      std::piecewise_construct, std::forward_as_tuple(name), std::forward_as_tuple(type, params));
+
+  // This should be a sub mesh generator. We can assume this because if we're in the middle of
+  // constructing mesh generators (not "adding" them, where we simply store their parameters),
+  // that we're constructing a sub generator also in the correct order
```

Comment:
```suggestion
  // we should be constructing a sub generator also in the correct order, as they are naturally chained
  // in the mesh generator creating them.
```

Reply:
They're not naturally chained.

Reply:
grammar seems to not work. "we can assume this because if .... that something"

you can replace naturally chain with something that is more simple